File: project-euler/c/prob003.c
#include <math.h>
#include <stdio.h>

long largest_prime(double n) {
    for (double i = ceil(sqrt(n)); i > 1; i -= 1.0)
        if (fmod(n, i) == 0.0 && largest_prime(i) == 1)
            return (long)i;
    return 1;
}

int main() {
    double num = 600851475143.0;
    long factor = largest_prime(num);
    printf("factor: %ld\n", factor);
}

File: project-euler/lua/prob003.lua
local common = require("common")

---@type function
local factor

---@param n integer
factor = function(n)
    local tab = {}
    if n <= 1 then return tab end
    for i = 2, math.ceil(math.sqrt(n)) do
        if n % i == 0 then
            tab[i] = true
            for val in pairs(factor(n / i)) do
                if not tab[val] then
                    tab[val] = true
                end
            end
            print("factors: " .. table.concat(tab, " "))
        end
    end
    return tab
end

local main = function()
    local n = 600851475143
    print(factor(n))
end

main()